File: frontend/src/utils/frontmatterUtils.js
/**
 * Frontmatter Parsing Utilities - Roosevelt's Shared Cavalry Tools
 * 
 * **BULLY!** Consistent frontmatter parsing across the application!
 */

/**
 * Parse YAML-like frontmatter from markdown text
 * **By George!** Extract metadata like a well-organized cavalry charge!
 */
export function parseFrontmatter(text) {
  try {
    const trimmed = text.startsWith('\ufeff') ? text.slice(1) : text;
    if (!trimmed.startsWith('---\n')) return { data: {}, lists: {}, order: [], raw: '', body: text };
    const end = trimmed.indexOf('\n---', 4);
    if (end === -1) return { data: {}, lists: {}, order: [], raw: '', body: text };
    const yaml = trimmed.slice(4, end).replace(/\r/g, '');
    const body = trimmed.slice(end + 4).replace(/^\n/, '');
    const data = {};
    const lists = {};
    const order = [];
    const lines = yaml.split('\n');
    for (let i = 0; i < lines.length; i++) {
      const line = lines[i];
      const m = line.match(/^([A-Za-z0-9_\-]+):\s*(.*)$/);
      if (m) {
        const k = m[1].trim();
        const v = m[2];
        order.push(k);
        if (v && v.trim().length > 0) {
          data[k] = String(v).trim();
        } else {
          const items = [];
          let j = i + 1;
          while (j < lines.length) {
            const ln = lines[j];
            if (/^\s*-\s+/.test(ln)) {
              items.push(ln.replace(/^\s*-\s+/, ''));
              j++;
            } else if (/^\s+$/.test(ln)) {
              j++;
            } else {
              break;
            }
          }
          if (items.length > 0) {
            lists[k] = items;
            i = j - 1;
          } else {
            data[k] = '';
          }
        }
      }
    }
    return { data, lists, order, raw: yaml, body };
  } catch (e) {
    return { data: {}, lists: {}, order: [], raw: '', body: text };
  }
}

/**
 * Build frontmatter YAML block from data object
 * **BULLY!** Construct metadata like a well-drilled regiment!
 */
export function buildFrontmatter(data) {
  if (!data || typeof data !== 'object') return '';
  
  const lines = [];
  for (const [key, value] of Object.entries(data)) {
    if (value !== null && value !== undefined) {
      lines.push(`${key}: ${value}`);
    }
  }
  
  return lines.length > 0 ? `---\n${lines.join('\n')}\n---\n` : '';
}

/**
 * Check if content has frontmatter
 * **By George!** Quick reconnaissance of document structure!
 */
export function hasFrontmatter(text) {
  if (!text || typeof text !== 'string') return false;
  const trimmed = text.startsWith('\ufeff') ? text.slice(1) : text;
  return trimmed.startsWith('---\n') && trimmed.indexOf('\n---', 4) !== -1;
}

/**
 * Extract just the frontmatter data without parsing the full structure
 * **BULLY!** Quick extraction for lightweight operations!
 */
export function extractFrontmatterData(text) {
  const parsed = parseFrontmatter(text);
  return parsed.data || {};
}

File: backend/utils/frontmatter_utils.py
"""
Minimal YAML-like frontmatter parser for Markdown files.

This utility mirrors the lightweight parsing used in the frontend
and supports simple key: value lines. It ignores complex YAML constructs
to avoid extra dependencies.
"""

from typing import Dict, Tuple


def parse_frontmatter(text: str) -> Tuple[Dict[str, str], str]:
    """
    Parse a simple YAML-like frontmatter block from the top of the text.

    Returns (data, body). Tolerant of \r\n or \n newlines and optional BOM.
    """
    try:
        if not text:
            return {}, ""

        import re

        trimmed = text[1:] if text.startswith("\ufeff") else text
        m = re.match(r"^---\s*\r?\n([\s\S]*?)\r?\n---\s*\r?\n?", trimmed)
        if not m:
            return {}, text

        yaml_block = m.group(1)
        body = trimmed[m.end():]

        data: Dict[str, str] = {}
        for line in re.split(r"\r?\n", yaml_block):
            line = line.strip()
            if not line or line.startswith("#"):
                continue
            if ":" in line:
                k, v = line.split(":", 1)
                data[k.strip()] = v.strip()

        return data, body
    except Exception:
        return {}, text


def build_frontmatter(data: Dict[str, str]) -> str:
    lines = []
    for k, v in data.items():
        if v is None:
            continue
        vs = str(v)
        if len(vs) == 0:
            continue
        lines.append(f"{k}: {vs}")
    if not lines:
        return ""
    return "---\n" + "\n".join(lines) + "\n---\n"




